File: MotoActions/src/org/lineageos/settings/device/DozePreferenceFragment.java
/*
 * SPDX-FileCopyrightText: 2015 The CyanogenMod Project
 * SPDX-FileCopyrightText: The LineageOS Project
 * SPDX-License-Identifier: Apache-2.0
 */

package org.lineageos.settings.device;

import android.app.AlertDialog;
import android.app.Dialog;
import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;

import androidx.preference.Preference;
import androidx.preference.PreferenceCategory;
import androidx.preference.PreferenceFragmentCompat;
import androidx.preference.SwitchPreferenceCompat;

import com.android.settingslib.widget.MainSwitchPreference;

public class DozePreferenceFragment extends PreferenceFragmentCompat
        implements Preference.OnPreferenceChangeListener {

    private SwitchPreferenceCompat mAlwaysOnDisplayPreference;

    private SwitchPreferenceCompat mHandwavePreference;
    private SwitchPreferenceCompat mPickUpPreference;
    private SwitchPreferenceCompat mPocketPreference;

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        setPreferencesFromResource(R.xml.doze_panel, rootKey);

        SharedPreferences prefs =
                getActivity().getSharedPreferences("doze_panel", Context.MODE_PRIVATE);
        if (savedInstanceState == null && !prefs.getBoolean("first_help_shown", false)) {
            showHelp();
        }

        boolean dozeEnabled = MotoActionsSettings.isDozeEnabled(getActivity());

        MainSwitchPreference switchBar = findPreference(MotoActionsSettings.DOZE_ENABLE);
        switchBar.setOnPreferenceChangeListener(this);
        switchBar.setChecked(dozeEnabled);

        mAlwaysOnDisplayPreference = findPreference(MotoActionsSettings.ALWAYS_ON_DISPLAY);
        mAlwaysOnDisplayPreference.setEnabled(dozeEnabled);
        mAlwaysOnDisplayPreference.setChecked(MotoActionsSettings.isAlwaysOnEnabled(getActivity()));
        mAlwaysOnDisplayPreference.setOnPreferenceChangeListener(this);

        mHandwavePreference = findPreference(MotoActionsSettings.GESTURE_IR_WAKEUP_KEY);
        mHandwavePreference.setEnabled(dozeEnabled);
        mHandwavePreference.setOnPreferenceChangeListener(this);

        mPickUpPreference = findPreference(MotoActionsSettings.GESTURE_PICK_UP_KEY);
        mPickUpPreference.setEnabled(dozeEnabled);
        mPickUpPreference.setOnPreferenceChangeListener(this);

        mPocketPreference = findPreference(MotoActionsSettings.GESTURE_POCKET_KEY);
        mPocketPreference.setEnabled(dozeEnabled);
        mPocketPreference.setOnPreferenceChangeListener(this);

        // Hide AOD if not supported and set all its dependents otherwise
        if (!MotoActionsSettings.alwaysOnDisplayAvailable(getActivity())) {
            getPreferenceScreen().removePreference(mAlwaysOnDisplayPreference);
        } else {
            PreferenceCategory ambientDisplayCategory = findPreference("ambient_display_key");
            ambientDisplayCategory.setDependency(MotoActionsSettings.ALWAYS_ON_DISPLAY);
        }
    }

    @Override
    public boolean onPreferenceChange(Preference preference, Object newValue) {
        boolean isChecked = (Boolean) newValue;
        if (MotoActionsSettings.ALWAYS_ON_DISPLAY.equals(preference.getKey())) {
            MotoActionsSettings.enableAlwaysOn(getActivity(), isChecked);
        } else if (MotoActionsSettings.DOZE_ENABLE.equals(preference.getKey())) {
            MotoActionsSettings.enableDoze(getActivity(), isChecked);

            if (!isChecked) {
                MotoActionsSettings.enableAlwaysOn(getActivity(), false);
                mAlwaysOnDisplayPreference.setChecked(false);
            }
            mAlwaysOnDisplayPreference.setEnabled(isChecked);

            mHandwavePreference.setEnabled(isChecked);
            mPickUpPreference.setEnabled(isChecked);
            mPocketPreference.setEnabled(isChecked);
        }

        return true;
    }

    private void showHelp() {
        AlertDialog helpDialog = new AlertDialog.Builder(getActivity())
                .setTitle(R.string.doze_settings_help_title)
                .setMessage(R.string.doze_settings_help_text)
                .setPositiveButton(R.string.dialog_ok,
                        (dialog, which) -> {
                            getActivity()
                                    .getSharedPreferences("doze_panel", Context.MODE_PRIVATE)
                                    .edit()
                                    .putBoolean("first_help_shown", true)
                                    .commit();
                            dialog.cancel();
                        })
                .create();
        helpDialog.show();
    }
}

File: packages/MotoActions/src/org/lineageos/settings/device/DozePreferenceFragment.kt
/*
 * SPDX-FileCopyrightText: 2015 The CyanogenMod Project
 * SPDX-FileCopyrightText: The LineageOS Project
 * SPDX-License-Identifier: Apache-2.0
 */

package org.lineageos.settings.device

import android.content.Context
import android.os.Bundle
import androidx.appcompat.app.AlertDialog
import androidx.preference.Preference
import androidx.preference.PreferenceCategory
import androidx.preference.SwitchPreferenceCompat
import com.android.settingslib.widget.MainSwitchPreference
import com.android.settingslib.widget.SettingsBasePreferenceFragment

class DozePreferenceFragment :
    SettingsBasePreferenceFragment(), Preference.OnPreferenceChangeListener {

    private lateinit var alwaysOnDisplayPreference: SwitchPreferenceCompat
    private lateinit var handwavePreference: SwitchPreferenceCompat
    private lateinit var pickUpPreference: SwitchPreferenceCompat
    private lateinit var pocketPreference: SwitchPreferenceCompat

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        setPreferencesFromResource(R.xml.doze_panel, rootKey)

        val prefs = requireActivity().getSharedPreferences("doze_panel", Context.MODE_PRIVATE)

        if (savedInstanceState == null && !prefs.getBoolean("first_help_shown", false)) {
            showHelp()
        }

        val dozeEnabled = MotoActionsSettings.isDozeEnabled(requireActivity())

        val switchBar: MainSwitchPreference = findPreference(MotoActionsSettings.DOZE_ENABLE)!!
        switchBar.onPreferenceChangeListener = this
        switchBar.isChecked = dozeEnabled

        alwaysOnDisplayPreference = findPreference(MotoActionsSettings.ALWAYS_ON_DISPLAY)!!
        alwaysOnDisplayPreference.isEnabled = dozeEnabled
        alwaysOnDisplayPreference.isChecked =
            MotoActionsSettings.isAlwaysOnEnabled(requireActivity())
        alwaysOnDisplayPreference.onPreferenceChangeListener = this

        handwavePreference = findPreference(MotoActionsSettings.GESTURE_IR_WAKEUP_KEY)!!
        handwavePreference.isEnabled = dozeEnabled
        handwavePreference.onPreferenceChangeListener = this

        pickUpPreference = findPreference(MotoActionsSettings.GESTURE_PICK_UP_KEY)!!
        pickUpPreference.isEnabled = dozeEnabled
        pickUpPreference.onPreferenceChangeListener = this

        pocketPreference = findPreference(MotoActionsSettings.GESTURE_POCKET_KEY)!!
        pocketPreference.isEnabled = dozeEnabled
        pocketPreference.onPreferenceChangeListener = this

        if (!MotoActionsSettings.alwaysOnDisplayAvailable(requireActivity())) {
            preferenceScreen.removePreference(alwaysOnDisplayPreference)
        } else {
            val ambientDisplayCategory: PreferenceCategory = findPreference("ambient_display_key")!!
            ambientDisplayCategory.dependency = MotoActionsSettings.ALWAYS_ON_DISPLAY
        }
    }

    override fun onPreferenceChange(preference: Preference, newValue: Any): Boolean {
        val isChecked = newValue as Boolean

        when (preference.key) {
            MotoActionsSettings.ALWAYS_ON_DISPLAY -> {
                MotoActionsSettings.enableAlwaysOn(requireActivity(), isChecked)
            }

            MotoActionsSettings.DOZE_ENABLE -> {
                MotoActionsSettings.enableDoze(requireActivity(), isChecked)

                if (!isChecked) {
                    MotoActionsSettings.enableAlwaysOn(requireActivity(), false)
                    alwaysOnDisplayPreference.isChecked = false
                }

                alwaysOnDisplayPreference.isEnabled = isChecked
                handwavePreference.isEnabled = isChecked
                pickUpPreference.isEnabled = isChecked
                pocketPreference.isEnabled = isChecked
            }
        }

        return true
    }

    private fun showHelp() {
        AlertDialog.Builder(requireActivity())
            .setTitle(R.string.doze_settings_help_title)
            .setMessage(R.string.doze_settings_help_text)
            .setPositiveButton(R.string.dialog_ok) { dialog, _ ->
                requireActivity()
                    .getSharedPreferences("doze_panel", Context.MODE_PRIVATE)
                    .edit()
                    .putBoolean("first_help_shown", true)
                    .commit()
                dialog.cancel()
            }
            .create()
            .show()
    }
}
